requestbatcher,intentresolver: add timeout in requestbatcher and adopt

ajwerner opened the pull request:
This PR adds a configuration option for a timeout for batch request sends in
the request batcher and adopts this option in the intent resolver. Prior to
the introduction of batching in #34803 intent resolution was performed with a
maximum timeout of 30s. This PR uses that same constant and additionally
applies it to batches of requests to gc transaction records.

Informs #36806.

Release note: None

Commit:
requestbatcher,intentresolver: add timeout in requestbatcher and adopt

This PR adds a configuration option for a timeout for batch request sends in
the request batcher and adopts this option in the intent resolver. Prior to
the introduction of batching in #34803 intent resolution was performed with a
maximum timeout of 30s. This PR uses that same constant and additionally
applies it to batches of requests to gc transaction records.

Informs #36806.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36845)
<!-- Reviewable:end -->

ajwerner commented:
Before #34803 we used to also respect client provided timeouts via the context passed to the intent resolver if it was shorter than the package default. We could add logic to do that still but I wasn't sure what the right policy would be. It seems like if the code were to give any consideration to client provided timeouts the two options would be to respect the earliest timeout of any request sent in the batch or the latest. The earliest worries me as it's probably not a good idea to cancel a whole batch for a single request due perhaps to a timeout from processing the gc queue. Taking the latest client provided timeout seems fine but I don't have a compelling argument to justify the complexity.

Commit:
contextutil: return TimeoutError if context has DeadlineExceeded

Before this commit, if the context were to have expired but the returned
errors did not have a Cause of context.DeadlineExceeded, RunWithTimeout
would not return a TimeoutError.

It is perhaps surprising that the returned TimeoutError does not have
context.DeadlineExceeded as its Cause but it seemed preferable to
retain the potentially more meaningful error as the Cause.

ajwerner commented:
Added another commit which changes the behavior of contextutil.RunWithTimeout so that it returns a TimeoutError if the timeout is exceeded even if the returned error does not have context.DeadlineExceeded as its Cause. See the commit message for more details.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36845#-:-LcYKOvd0IXszKBij2V7:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/internal/client/requestbatcher/batcher.go, line 272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36845#-LcYJd_OESinddzgwBV7:-LcYJd_OESinddzgwBV8:b-ol3c5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/8de9061065a5e39a9e74704a43dc9b93de19a3cb/pkg/internal/client/requestbatcher/batcher.go#L272)):*
> ```Go
> 		defer b.sendDone(ba)
> 		var resp *roachpb.BatchResponse
> 		var pErr *roachpb.Error
> ```

Are we getting anything out of capturing `pErr`? If not then I'd tighten its scope both for clarity and to ensure that it doesn't escape.
___
*[pkg/util/contextutil/context_test.go, line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36845#-LcYIyIQ45BWfW4EKWFx:-LcYIyIQ45BWfW4EKWFy:b-qu54m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0042c7d5e8e11b663cc29e2435543b22282503aa/pkg/util/contextutil/context_test.go#L83)):*
> ```Go
> 	notContextDeadlineExceeded := errors.New(context.DeadlineExceeded.Error())
> 	err := RunWithTimeout(ctx, "foo", 1, func(ctx context.Context) error {
> 		time.Sleep(10 * time.Millisecond)
> ```

This looks flaky. Can't we wait on the context cancelation like: `<-ctx.Done()`?


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
> The earliest worries me as it's probably not a good idea to cancel a whole batch for a single request due perhaps to a timeout from processing the gc queue. Taking the latest client provided timeout seems fine but I don't have a compelling argument to justify the complexity.

I agree that it's not worth the complexity. What you have here looks fine.

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36845)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/internal/client/requestbatcher/batcher.go, line 272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36845#-LcYJd_OESinddzgwBV7:-LcYl08ICqHKeAfzcJrw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8de9061065a5e39a9e74704a43dc9b93de19a3cb/pkg/internal/client/requestbatcher/batcher.go#L272)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are we getting anything out of capturing `pErr`? If not then I'd tighten its scope both for clarity and to ensure that it doesn't escape.
</blockquote></details>

Done.
___
*[pkg/util/contextutil/context_test.go, line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36845#-LcYIyIQ45BWfW4EKWFx:-LcYkm0_2vAdjJ4P3jF4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0042c7d5e8e11b663cc29e2435543b22282503aa/pkg/util/contextutil/context_test.go#L83)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This looks flaky. Can't we wait on the context cancelation like: `<-ctx.Done()`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
